### PR TITLE
Part7

### PR DESCRIPTION
**Pull request by MasonSelf**

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -272,406 +222,224 @@ struct HeapA
...
-        FloatType& pow( const DoubleType& dt ); 
-        
-        FloatType& apply( std::function<FloatType&( float& )> f );
-        FloatType& apply( void( *f )( float& ) );
+        Numeric& operator+= ( Type mod )
+        {
+        if ( value != nullptr )
+            *value += mod;
+        return *this;
+        }
```
Comment (matkatmusic) [inline main.cpp:252]: Fix this indentation
- [ ] revise

File: main.cpp [hunk 2]
```diff
@@ -272,406 +222,224 @@ struct HeapA
...
+                }
+
+                else if ( mod <  std::numeric_limits<DivType>::epsilon())
+                {
+                    std::cout << "can't divide integers by zero!" << std::endl;
+                    return *this;
+                }   
+            }
+
+            else if ( mod <  std::numeric_limits<Type>::epsilon())
```
Comment (matkatmusic) [inline main.cpp:288]: This should be `<DivType>` not `<Type>`
you're comparing if your parameter is less than its type's epsilon, not your class type's epsilon.
Also, if `Type` has an epsilon of 0.f, and `mod` is -1.f, will this `if()` block be entered?
- [ ] revise

File: main.cpp [hunk 3]
```diff
@@ -272,406 +222,224 @@ struct HeapA
...
-{
-    if (value != nullptr )
-        *value *= mod;
-    return *this;
-}
+        Numeric& operator+= ( Type mod )
+        {
+        if ( value != nullptr )
+            *value += mod;
+        return *this;
```
Comment (matkatmusic) [inline main.cpp:343]: Fix this indentation

- [ ] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: main.cpp [hunk 4]
```diff
@@ -272,406 +222,224 @@ struct HeapA
...
+                }
+
+                else if ( mod <  std::numeric_limits<DivType>::epsilon())
+                {
+                    std::cout << "can't divide integers by zero!" << std::endl;
+                    return *this;
+                }   
+            }
+
+            else if ( mod <  std::numeric_limits<DivType>::epsilon())
```
Comment (matkatmusic) [inline main.cpp:288]: This doesn't answer my question about: if the `epsilon` of `DivType` is `0.f`, and `mod is `-1.f`, is this `if` block entered?  and if it is, is the `std::cout` statement correct?
- [ ] revise

File: main.cpp [hunk 5]
```diff
@@ -272,406 +222,224 @@ struct HeapA
...
+                if constexpr (std::is_same<DivType, int>::value)
+                {
+                    if ( mod == 0 )
+                    {
+                        std::cout << "error: integer division by zero is an error and will crash the program!" << std::endl;
+                        return *this;
+                    }
+                }
+
+                else if ( mod <  std::numeric_limits<DivType>::epsilon())
```
Comment (matkatmusic) [inline main.cpp:281]: This doesn't answer my question about: if the `epsilon` of `DivType` is `0.f`, and `mod is `-1.f`, is this `if` block entered?  and if it is, is the `std::cout` statement correct?
- [ ] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: main.cpp (hunk 4 above)
Comment (MasonSelf) [reply main.cpp]: abs will account for that

File: main.cpp (hunk 5 above)
Comment (MasonSelf) [reply main.cpp]: abs will account for that

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**